chain/ethereum/src/network.rs

```diff
@@ -48,64 +72,129 @@ impl EthereumNetworkAdapter {
+            retest_percent: DEFAULT_ADAPTER_ERROR_RETEST_PERCENT,
         }
     }
 
-    pub fn push_adapter(&mut self, adapter: EthereumNetworkAdapter) {
-        if adapter.is_call_only() {
-            self.call_only_adapters.push(adapter);
-        } else {
-            self.adapters.push(adapter);
+    #[cfg(debug_assertions)]
+    pub async fn for_testing(
+        mut adapters: Vec<EthereumNetworkAdapter>,
+        call_only: Vec<EthereumNetworkAdapter>,
+    ) -> Self {
+        use std::cmp::Ordering;
+
+        use graph::slog::{o, Discard, Logger};
+
+        use graph::components::adapter::MockIdentValidator;
+        let chain_id: ChainId = "testing".into();
+        adapters.sort_by(|a, b| {
+            a.capabilities
+                .partial_cmp(&b.capabilities)
+                .unwrap_or(Ordering::Equal)
+        });
+
+        let provider = ProviderManager::new(
+            Logger::root(Discard, o!()),
+            vec![(chain_id.clone(), adapters)].into_iter(),
+            Arc::new(MockIdentValidator),
+        );
+        provider.mark_all_valid().await;
+
+        Self::new(chain_id, provider, call_only, None)
+    }
+
+    pub fn new(
+        chain_id: ChainId,
+        manager: ProviderManager<EthereumNetworkAdapter>,
+        call_only_adapters: Vec<EthereumNetworkAdapter>,
+        retest_percent: Option<f64>,
+    ) -> Self {
+        #[cfg(debug_assertions)]
+        call_only_adapters.iter().for_each(|a| {
+            a.is_call_only();
+        });
+
+        Self {
+            chain_id,
+            manager,
+            call_only_adapters,
+            retest_percent: retest_percent.unwrap_or(DEFAULT_ADAPTER_ERROR_RETEST_PERCENT),
         }
     }
-    pub fn all_cheapest_with(
-        &self,
+
+    fn available_with_capabilities<'a>(
+        input: Vec<&'a EthereumNetworkAdapter>,
         required_capabilities: &NodeCapabilities,
-    ) -> impl Iterator<Item = &EthereumNetworkAdapter> + '_ {
-        let cheapest_sufficient_capability = self
-            .adapters
+    ) -> impl Iterator<Item = &'a EthereumNetworkAdapter> + 'a {
+        let cheapest_sufficient_capability = input
             .iter()
             .find(|adapter| &adapter.capabilities >= required_capabilities)
             .map(|adapter| &adapter.capabilities);
 
-        self.adapters
-            .iter()
+        input
+            .into_iter()
             .filter(move |adapter| Some(&adapter.capabilities) == cheapest_sufficient_capability)
             .filter(|adapter| adapter.get_capacity() > AvailableCapacity::Unavailable)
     }
 
-    pub fn cheapest_with(
+    pub async fn all_cheapest_with(
         &self,
         required_capabilities: &NodeCapabilities,
+    ) -> impl Iterator<Item = &EthereumNetworkAdapter> + '_ {
+        let all = self
+            .manager
+            .get_all(&self.chain_id)
+            .await
+            .unwrap_or_default();
```

Comment:
Is the intent behind the `.unwrap_or_default()` that we'll return an empty iterator?  It might be better if this returned `Result<impl Iterator<Item = &EthereumNetworkAdapter> + '_, SomeError>` to force the caller to handle that condition. If not, the method should at least have a comment on the circumstances under which this returns an empty iterator.

Reply:
added a comment